modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js

```diff
@@ -0,0 +1,272 @@
+  depthLimit: Number.MAX_SAFE_INTEGER,
+  onTilesetLoad: tileset3d => {},
+  onTileLoad: tileHeader => {}
+};
+
+export default class Tile3DLayer extends CompositeLayer {
+  initializeState() {
+    this.state = {
+      layerMap: {},
+      layers: [],
+      tileset3d: null
+    };
+  }
+
+  shouldUpdateState({changeFlags}) {
+    return changeFlags.somethingChanged;
+  }
+
+  async updateState({props, oldProps}) {
+    if (props.tilesetUrl && props.tilesetUrl !== oldProps.tilesetUrl) {
+      await this._loadTileset(props.tilesetUrl);
+    } else if (
+      (props.ionAccessToken || props.ionAssetId) &&
+      (props.ionAccessToken !== oldProps.ionAccessToken || props.ionAssetId !== oldProps.ionAssetId)
+    ) {
+      await this._loadTilesetFromIon(props.ionAccessToken, props.ionAssetId);
+    }
+
+    const {tileset3d} = this.state;
+    this._updateTileset(tileset3d);
+  }
+
+  async _loadTileset(tilesetUrl, fetchOptions, ionMetadata) {
+    let tileset3d = null;
+
+    if (tilesetUrl) {
+      const response = await fetch(tilesetUrl, fetchOptions);
+      const tilesetJson = await response.json();
+
+      tileset3d = new Tileset3D(tilesetJson, tilesetUrl, {
+        throttleRequests: true,
+        onTileLoad: tileHeader => {
+          this.props.onTileLoad(tileHeader);
+          this._updateTileset(tileset3d);
+        },
+        // TODO: explicit passing should not be needed, registerLoaders should suffice
+        DracoLoader: this.props.DracoWorkerLoader || this.props.DracoLoader,
+        fetchOptions,
+        ...ionMetadata
+      });
+    }
+
+    this.setState({
+      tileset3d,
+      layerMap: {},
+      layers: []
+    });
+
+    if (tileset3d) {
+      // TODO: Remove these after sse traversal is working since this is just to prevent full load of tileset and loading of root
+      tileset3d.depthLimit = this.props.depthLimit;
+      this.props.onTilesetLoad(tileset3d);
+    }
+  }
+
+  async _loadTilesetFromIon(ionAccessToken, ionAssetId) {
+    const ionMetadata = await _getIonTilesetMetadata(ionAccessToken, ionAssetId);
+    const {url, headers} = ionMetadata;
+    return await this._loadTileset(url, {headers}, ionMetadata);
+  }
+
+  _updateTileset(tileset3d) {
+    const {animationProps, viewport} = this.context;
+    if (!animationProps || !viewport || !tileset3d) {
+      return;
+    }
+
+    const frameState = getFrameState(viewport, animationProps);
+    tileset3d.update(frameState);
+    this._updateLayerMap();
+    this._selectLayers(frameState.frameNumber);
+  }
+
+  // `Layer` instances is created and added to the map if it doesn't exist yet.
+  _updateLayerMap() {
+    const {tileset3d, layerMap} = this.state;
+    const {selectedTiles} = tileset3d;
+
+    const tilesWithoutLayer = selectedTiles.filter(tile => !layerMap[tile.contentUri]);
+
+    for (const tile of tilesWithoutLayer) {
+      // TODO - why do we call this here? Being "selected" should automatically add it to cache?
```

Comment:
@loshjawrence - Do you think we can drop this line now?

Reply:
if you want to move the line to tileset.update after traversal that's fine.